corehq/apps/sms/views.py

```diff
@@ -787,129 +790,183 @@ def _fmt(d):
-                sender = doc.name
-            else:
-                sender = doc.first_name or doc.raw_username
-        elif sms.chat_user_id is not None:
-            if sms.chat_user_id in username_map:
-                sender = username_map[sms.chat_user_id]
+    @method_decorator(require_permission(Permissions.edit_data))
+    @method_decorator(requires_privilege_with_fallback(privileges.OUTBOUND_SMS))
+    def dispatch(self, request, *args, **kwargs):
+        return super(ChatMessageHistory, self).dispatch(request, *args, **kwargs)
+
+    @property
+    @memoized
+    def contact_id(self):
+        return self.request.GET.get('contact_id')
+
+    @property
+    @memoized
+    def contact(self):
+        if not self.contact_id:
+            return None
+
+        contact = get_contact(self.contact_id)
+        if not contact or contact.domain != self.domain:
+            return None
+
+        return contact
+
+    @property
+    @memoized
+    def contact_name(self):
+        if self.contact.doc_type == 'CommCareCase':
+            if self.domain_object.custom_case_username:
+                return self.contact.get_case_property(self.domain_object.custom_case_username)
             else:
-                try:
-                    user = CouchUser.get_by_user_id(sms.chat_user_id)
-                    sender = user.first_name or user.raw_username
-                except Exception:
-                    sender = _("Unknown")
-                username_map[sms.chat_user_id] = sender
+                return self.contact.name
         else:
-            sender = _("System")
-        last_sms = sms
-        result.append({
-            "sender": sender,
-            "text": sms.text,
-            "timestamp": (
-                ServerTime(sms.date).user_time(timezone)
-                .ui_string("%I:%M%p %m/%d/%y").lower()
-            ),
-            "utc_timestamp": json_format_datetime(sms.date),
-            "sent_by_requester": (sms.chat_user_id == request.couch_user.get_id),
-        })
-    if last_sms:
+            return self.contact.first_name or self.contact.raw_username
+
+    @quickcache(['user_id'], timeout=60 * 60, memoize_timeout=5 * 60)
+    def get_chat_user_name(self, user_id):
+        if not user_id:
+            return _("System")
+
         try:
-            entry, lock = LastReadMessage.get_locked_obj(
-                sms.domain,
-                request.couch_user._id,
-                sms.couch_recipient,
-                create=True
+            user = CouchUser.get_by_user_id(user_id)
+            return user.first_name or user.raw_username
+        except:
+            return _("Unknown")
+
+    @property
+    @memoized
+    def start_date_str(self):
+        return self.request.GET.get('start_date')
+
+    @property
+    @memoized
+    def start_date(self):
+        if not self.start_date_str:
+            return None
+
+        try:
+            return iso_string_to_datetime(self.start_date_str)
+        except:
```

Comment:
nit: may want to make this just `ValueError`, but if it doesn't matter too much and only cares about succeeding then fine to leave